src/Aspire.Hosting/Publishing/AutomaticManifestPublisherBindingInjectionHook.cs

```diff
@@ -28,6 +29,15 @@ private static bool IsKestrelHttp2ConfigurationPresent(ProjectResource projectRe
         return protocol == "Http2";
     }
 
+    private static bool IsWebProject(ProjectResource projectResource)
+    {
+        var serviceMetadata = projectResource.GetServiceMetadata();
+        using var projectFileStream = File.OpenRead(serviceMetadata.ProjectPath);
+        var document = new XmlDocument();
+        document.Load(projectFileStream);
+        return document.DocumentElement?.GetAttribute("Sdk") == "Microsoft.NET.Sdk.Web";
```

Comment:
@baronfel

Reply:
This will work for the common case, but as @mitchdenny mentions below the following is currently allowed, for example:

```xml
<Project Sdk="Microsoft.NET.Sdk.Web;Microsoft.NET.Sdk.Publish">
```

_also_ you can declare SDKs in the project in a decoupled format:

```xml

<Project>
  
  <!-- some logic you want to happen before props evaluation of the SDK -->
  <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
  <!-- 'normal' project file contents -->
  <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />
  <!-- some logic you want to happen after targets evaluation of the SDK -->
</Project>
```

You're really looking at MSBuild evaluations here if you want a _complete_ answer. Most SDKs set some properties that declare capabilities, it might be easiest to do an MSBuild evaluation of the project and then ask "does this capability exist on the project". This is very similar to how the IDEs light up features.

Reply:
We want to use our current msbuild logic to extract this data and make it part of our codegen.

Reply:
I'm looking through the structured log viewer trying to see if there is anyway to figure this out. So far I'm coming up blank. Hopefully @baronfel can shed some light.

Reply:
@mitchdenny would a property set by the Web SDK be good enough? if so you can check for [UsingMicrosoftNETSdkWeb](https://github.com/dotnet/sdk/blob/846338911ef8f8d18f0bc745b8108a812ca4c7a1/src/WebSdk/Web/Targets/Sdk.Server.props#L17).

Reply:
I was just looking at that: `dotnet msbuild --getProperty:UsingMicrosoftNETSdkWeb`

But I need to get that for each project reference from within out targets. Is there a way to do this without having to invoke msbuild on each project reference (seems expensive).

Reply:
You need to get MSBuild involved. We can't be parsing the XML and trying to interpret properties ourselves. That is a pit of failure.

Reply:
> but @mitchdenny is saying that we don't know how to get this information out of the project reference.

You need to invoke MSBuild on the project reference. It is the only way to get this information reliably.

Maybe we can store this information somewhere? like on the `IServiceMetadata` interface, where we have the `AssemblyPath` already. From the AppHost project, we should be able to call a Target on the referenced project to get if it is a web project or not. Then we shove that info into the generated class.

Maybe?

Reply:
It would be idea if it happened as part of the project build so we don't need to invoke msbuild at runtime from the apphost.

Reply:
Here's a concrete example of what that would look like, from an experiment I did around moving the host towards using the ProjectReferences explicitly instead of using more second-order data points like `ReferencePathWithRefAssemblies`:

https://github.com/baronfel/aspire/commit/d1aef730e538dbb29be96ac1ae0c8e6bf3fcf279#diff-799cc1aaddb026479eeb944711fde0753d2f783b01903de3cc741998566acd87R20-R22

Reply:
Can we do this right now and see how bad the perf is? eShop would be a place to verify if we're destroying the performance or not 😄

Reply:
perf of what? adding this to the manifest generation target? I wouldn't expect this to move the needle in terms of overall speed of manifest generation at all - especially because that target still has a dependency on `Build`, which causes _all_ of the ProjectReferences to build.

Reply:
ah, yes. Still not a problem IMO. The imaginary target we're talking about being invoked here (if we're talking about ProjectCapabilities) is like returning a constant - all of the data is computed during evaluation, which has already happened. so you _should_ just be paying MSBuild Target invocation cost, which is nonzero. Here's from that branch I linked for reference:

![image](https://github.com/dotnet/aspire/assets/573979/99cdf7ef-7db6-4566-a2a3-171934a3daf9)

28ms for 18 projects (yes most of these are internal libraries, that's one of the hurdles that I still have to figure out). That's for a target whose result is already cached, which I believe would/could be the case here.

Reply:
> Here's a concrete example of what that "call a target for each project and use the results" flow would look like, from an experiment I did around moving the host towards using the ProjectReferences explicitly instead of using more second-order data points like `ReferencePathWithRefAssemblies`:
> 
> [baronfel/aspire@d1aef73#diff-799cc1aaddb026479eeb944711fde0753d2f783b01903de3cc741998566acd87R20-R22](https://github.com/baronfel/aspire/commit/d1aef730e538dbb29be96ac1ae0c8e6bf3fcf279#diff-799cc1aaddb026479eeb944711fde0753d2f783b01903de3cc741998566acd87R20-R22)

I pasted this into the code base but its complaining about this:

```
C:\Code\dotnet\aspire\src\Aspire.Hosting\build\Aspire.Hosting.targets(43,9): error MSB4186: Invalid static method invocation syntax: "[System.IO.Path]::GetFullPath()". Method 'System.IO.Path.GetFullPath' not found. Static method invocation should be of the form: $([FullTypeName]::Method()), e.g. $([System.IO.Path]::Combine(`a`, `b`)). Check that all parameters are defined, are of the correct type, and are specified in the right order. [C:\Code\dotnet\aspire\samples\eShopLite\AppHost\AppHost.csproj]
```

Seems like %(Identity) isn't resolving. Did this actually work locally for you or is it pseudo code?

Reply:
Ah ... you filter out the project references which include references to assemblies.

Reply:
@baronfel I can't quite connect in my head how your example allows me to get at the `UsingMicrosoftNETSdkWeb` property value.